egs/librispeech/asr/simple_v1/bpe_ctc_att_conformer_train.py

```diff
@@ -0,0 +1,633 @@
+from snowfall.training.diagnostics import measure_gradient_norms, optim_step_and_measure_param_change
+
+
+# DDP seems disable logging as discussed here https://github.com/k2-fsa/snowfall/pull/158
+# Temporarily use print instead of logging.info
+logging.info = print
+
+def get_objf(batch: Dict,
+             model: AcousticModel,
+             device: torch.device,
+             numericalizer: Numericalizer,
+             is_training: bool,
+             is_update: bool,
+             accum_grad: int = 1,
+             att_rate: float = 0.0,
+             tb_writer: Optional[SummaryWriter] = None,
+             global_batch_idx_train: Optional[int] = None,
+             optimizer: Optional[torch.optim.Optimizer] = None,
+             ):
+    feature = batch['inputs']
+    # at entry, feature is [N, T, C]
+    feature = feature.permute(0, 2, 1)  # now feature is [N, C, T]
+    assert feature.ndim == 3
+    feature = feature.to(device)
+
+    supervisions = batch['supervisions']
+
+    supervision_segments = torch.stack(
+        (supervisions['sequence_idx'],
+         (((supervisions['start_frame'] - 1) // 2 - 1) // 2),
+         (((supervisions['num_frames'] - 1) // 2 - 1) // 2)), 1
+    ).to(torch.int32)
+
+    texts = supervisions['text']
+
+    grad_context = nullcontext if is_training else torch.no_grad
+    with grad_context():
+        # nnet_output is [N, C, T]
+        nnet_output, encoder_memory, memory_mask = model(feature, supervisions)
+
+        blank_id = 0
+        unk_id = 1 # i.e. oov_id
+        token_ids = []
+        for text in texts:
+            token_ids.append(list(filter(lambda x: x != blank_id and x != unk_id, numericalizer.EncodeAsIds(text.upper()))))
+
+
+        if att_rate != 0.0:
+            att_loss = model.module.decoder_forward(encoder_memory, memory_mask, token_ids=token_ids)
+
+        # Prepare to compute ctc_loss
+        nnet_output = nnet_output.permute(2, 0, 1)  # Now is [T, N, C], as required by torch.nn.CTCLoss
+
+        target_lengths = torch.tensor([len(token_id) for token_id in token_ids])  # N
+        target= torch.tensor(list(np.concatenate(token_ids).flat))  # size is sum(target_lengths)
+        assert sum(target_lengths) == len(target)
+
+        input_lengths = supervision_segments[:,2]
+
+        bni = input_lengths.shape[0]  # batch_size of input
+        bno = nnet_output.shape[1]  # batch_size of nnet_output
+        bnt = target_lengths.shape[0]  # batch_size of target_legnths
+        assert bno == bni and bno == bnt and bni == bnt, 'Inconsistent batch-size!'
+
+        ctc_loss = model.module.ctc_loss_fn(nnet_output, target, input_lengths, target_lengths)
+
+        # Normalized by batch_size
+        # Reference: https://github.com/espnet/espnet/blob/master/espnet2/asr/ctc.py#L98
+        ctc_loss = ctc_loss.sum() / bno
+
+        if att_rate != 0.0:
+            loss = ((1.0 - att_rate) * ctc_loss + att_rate * att_loss) *  accum_grad
```

Comment:
>As we mentioned in person, I believe a problem with the current setup is that the transformer loss is being normalized (divided by the minibatch size) twice, once in a library function and once in the training script, while the CTC loss is only normalized once.
If we had logged the 2 objectives separately, we likely would have noticed this.

@danpovey I don't think att_loss is normalized twice. After it is computed at line 85, there is no extra normalization for att_loss anymore. 

BTW, the reduplicated normization for att_loss your mentioned maybe about [another code](https://github.com/pkufool/snowfall/blob/93dc56ebb63f226fc399e19ec218c7ceeb42bbe2/egs/librispeech/asr/simple_v1/ctc_att_transformer_train.py#L98), which is 
```
        if att_rate != 0.0:
            loss = (- (1.0 - att_rate) * tot_score + att_rate * att_loss) / (len(texts) * accum_grad)
```

Reply:
Ah, OK.
In any case, IMO we shouldn't be normalizing even once.  (so remove the "/ bno" above; and remove the same thing in the library function that computes the att_loss).

IMO, we should also remove the gradient-clipping step; my feeling is that if it was helping before, it was helping because it was compensating for the normalization that shouldn't have been happening.  This setup is non-recurrent so gradient clipping should not be needed.  (However, if we encounter instabilities we can revisit this).

Reply:
The model that removing gradient-clipping and normalizing is training now. 
The loss curve is as follows, in the beginning of each epoch, the loss value increases suddenly. Did it encounter instabilities. 
|total loss|ctc loss|att loss|
|--|--|--|
|![image](https://user-images.githubusercontent.com/11765074/126928426-1a64ccdc-cc0f-426d-b397-1eb6a793ef5d.png)|![image](https://user-images.githubusercontent.com/11765074/126928251-5a621437-698f-440d-a4ac-a32eab6d671c.png)|![image](https://user-images.githubusercontent.com/11765074/126928465-b5059d20-1e85-4cd3-ae49-3805eef2bcd1.png)|

The displaying loss value in tensorboard is normalized by num of utterances.